File: ng2-dart/lib/service/auth.service.dart
import 'dart:async';
import 'dart:convert';
import 'dart:html';

import 'package:angular2/core.dart';
import 'package:angular2/router.dart';

import 'package:http/http.dart';

import '../model/user.dart';

@Injectable()
class AuthService {
	
	static const _restUrl = '/rest/auth/';

	bool signedIn;
	User currentUser;

	final Router _router;
	final Client _http;

	AuthService(this._http, this._router);

	signOut() {
		signedIn = false;
		window.localStorage.remove('creds');
		currentUser = null;
	}

	Future<User> signIn(User user) async {

		var authData = window.btoa("$user.username:$user.password");

		var headas ={ 'Content-Type' : 'application/x-www-form-urlencoded',
						'Authorization': 'Basic $authData' };
		try {
			final response = await _http.get(_restUrl, headers: headas) ;
			final user = _extractData(response)
			.map((value) => new User.fromJson(value));

			_router.navigate(['']);

			return user;
		} catch (e) {
			throw _handleError(e);
		}
	}

	dynamic _extractData(Response resp) => JSON.decode(resp.body)['data'];

	Exception _handleError(dynamic e) {
    	print(e); // for demo purposes only
    	return new Exception('Server error; cause: $e');
  	}
}

File: ng2-ts/src/app/service/auth.service.ts
import { Injectable, EventEmitter } from '@angular/core';
import { Headers, Http } from '@angular/http';

import { Observable } from 'rxjs';

import { User } from '../model/user';

@Injectable()
export class AuthService {
 
  private restUrl = '/rest/auth';
  private signedIn: boolean;

  public currentUser: User;
  public currentUserChanged = new EventEmitter();
  public authLoading = new EventEmitter();

  constructor(private http: Http ) { }
 
  signOut() {
    this.signedIn = false;
    localStorage.removeItem('creds');
    this.currentUser = null;
    this.currentUserChanged.emit(this.currentUser);
  }
 
  signIn(user: User) {

    this.authLoading.emit(true);

    let authData = btoa(user.username + ":" + user.password);

    let headers = new Headers({
      'Content-Type': 'application/x-www-form-urlencoded',
      'Authorization': 'Basic ' + authData
    });
    
    return this.http
      .get(this.restUrl, {
        headers: headers
      })
      .map((res:any) => {
        this.authLoading.emit(false);
        if (res.status == 200) {
          
          this.signedIn = true;
          localStorage.setItem('creds', authData);
          this.currentUser = res.json() as User;

          this.currentUserChanged.emit(this.currentUser);

          return this.currentUser;
        } else {
          return null;
        }
      });
   }

   isSignedIn() {
     return this.signedIn == true && this.currentUser != null && localStorage.getItem('creds') != null;
   }
}